widgets/linechart/linechart_test.go

```diff
@@ -1636,6 +1637,48 @@ func TestLineChartDraws(t *testing.T) {
 				testdraw.MustBrailleLine(bc, image.Point{10, 2}, image.Point{11, 0})
 				testbraille.MustCopyTo(bc, c)
 
+				testcanvas.MustApply(c, ft)
+				return ft
+			},
+		},
+		{
+			desc:   "custom Y-axis labels using a value formatter",
```

Comment:
When it comes to test coverage, we probably have a few more interesting cases. And one of them will be a bit more complicated because well unicode. We probably want to test:

1. A case where the formatter returns an empty string for the label.

2. A case where the formatter returns a string that is so long that it won't even fit on the canvas. Which the widget should handle by giving up in a graceful way. Hopefully it would just somehow hit this code which simply draws a character we are using to indicate "there isn't enough space":
https://github.com/mum4k/termdash/blob/a9515f27215c8c7f3c0df731c4e86103cb9cdb9d/widgets/linechart/linechart.go#L303-L305

3. Is the harder one. Previously we generated labels internally, so we had full control over them. We could guarantee they will only contain printable ASCII characters. Now the user could include any non-printable (control or space) character or any unicode character.

3a. **Non-printable characters** - open to opinions here. My gut tells me that we should document they aren't allowed and return an error from Draw() if a label would contain one. We can use the following function that Termdash already uses in a few locations to validate the label:
https://github.com/mum4k/termdash/blob/a9515f27215c8c7f3c0df731c4e86103cb9cdb9d/internal/wrap/wrap.go#L61-L81

3b. **Unicode** - the trick with unicode is that some unicode runes take more than one terminal cell to draw. Some examples are here:
https://github.com/mum4k/termdash/blob/a9515f27215c8c7f3c0df731c4e86103cb9cdb9d/internal/runewidth/runewidth_test.go#L46-L48

We should support these as it is a goal of Termdash to be Unicode ready. When these characters (or graphemes) are in a string, the strings length no longer equals its width on the screen. At the very least this code isn't going to work anymore:
https://github.com/mum4k/termdash/blob/a9515f27215c8c7f3c0df731c4e86103cb9cdb9d/widgets/linechart/internal/axes/axes.go#L125-L134

We will have to update it to use runewidth.StringWidth instead of len() when determining the width of a label. The addition of the test case might uncover other issues.

Reply:
After adding the different test cases with the current code, these are the results:

- 1: Handles ok because it returns an error on the Draw, I think that for now, this could be correct although we could have an option in the future to disable labels on the different axis and this should not error.
- 2: Doesn't seem to reach that code:

```
               got:
                              │⢸
                              │⢸
                              │⢸
            51.680000000000000│⢸
                              │⡇
                              │⡇
                              │⡇
                             0│⡇
                              └─
                               0
              want:
                 │           ⢀⠔⠁
                 │          ⡠⠊
                 │        ⢀⠎
            61.7s│      ⢀⠔⠁
                 │     ⡠⠊
                 │   ⢀⠎
                 │ ⢀⠔⠁
                ∅│⡠⠊
                 └──────────────
                  0            1
```
I need to fix that case.

- 3.a: Without any change it returns a draw error, so we are ok also on this case.
- 3.b: It handles ok without any change, Do you think that this covers this case or I should use other unicode characters?

```
               got:                                                   
                 │           ⢀⠔⠁                                        
                 │          ⡠⠊                                          
                 │        ⢀⠎                                            
             ab世│      ⢀⠔⠁                                             
                 │     ⡠⠊                                               
                 │   ⢀⠎                                                
                 │ ⢀⠔⠁                                                 
              abc│⡠⠊                                          
                 └──────────────                              
                  0            1                                     
              want:                                              
                 │           ⢀⠔⠁                              
                 │          ⡠⠊                                
                 │        ⢀⠎                                  
             ab世│      ⢀⠔⠁                                          
                 │     ⡠⠊                                     
                 │   ⢀⠎                                                 
                 │ ⢀⠔⠁                                                  
              abc│⡠⠊                                                    
                 └──────────────                                        
                  0            1 
```
Although it seems to work without any change should I change `len` to use `runewidth.StringWidth`, I don't have enough experience with the lib to know if the Unicode chars are being checked in another code path along the way?

Reply:
Thanks a lot @slok for investigating this in detail.

(1) and (2) seem ok, I agree with your conclusions for (1) and I think the trimming behavior is most likely the best answer we can have for (2).

(3a) Looking at the code, I did not find any explicit validation other than checking for '\n' in this call stack:
https://github.com/mum4k/termdash/blob/e6af456c61cd50c650ae80857a74763d1ab4e158/widgets/linechart/internal/axes/label.go#L119-L122
https://github.com/mum4k/termdash/blob/e6af456c61cd50c650ae80857a74763d1ab4e158/internal/alignfor/align.go#L93-L96

Can you try with a formatter that returns one of these?
```go
// Space characters.
return "\ta"
// Control characters
fmt.Sprintf("%ca", 0x007f)
```

Basically characters that this test mentions ad invalid:
https://github.com/mum4k/termdash/blob/master/internal/wrap/wrap_test.go

Maybe when testing this, the chosen character triggered some other error? The other question we should ask is if the error returned from Draw was descriptive enough to make it clear what went wrong.

(3b) Looks like we will need to use runewidth.StringWidth and maybe also some other changes. Looking at the outputs above - it looks like we do have the bug, the unicode character caused an overflow into the next cell which shows itself as a loss of one space. A bit subtle, the unicode characters are tricky. Could you try with a string that will only have the wide characters? That should make the bug much more obvious as we will "lose" multiple cells.
```go
return "世世世世世世"
```